feature/dryformat: Refactor changelog generation

j4zz opened the pull request:
# Changes

* refactor: Moved changelog formatting into `lib/changelog.rb`
* refactor: Added various helper methods to make it easier to change formatting output and to make it less error-prone to change displayed information across multiple formats
* refactor: Change `Changelog#changes` to return hash keys `gui` and `refactoring` instead of `gui_changes` and `refactorings`
* fix: During changelog generation, use `commit_to` and `tag_to` instead of `commit_from` and `tag_from` to make an execution like `git-changelog 0.4.0 --format=slack` display information about the version being currently released
* feat: Got us started with a basic rspec setup and some test for the most complicated new methods in `lib/changelog.rb`
* feat: Add basic .travis.yml file to be able to start with CI

# Testing

* Please especially check the `commit_to` and `tag_to` change, since I’m not perfectly confident that this change is okay for all ways to execute `git-changelog`.
* You can run `rspec spec/` now. :)

j4zz commented:
This was the DRYing up part mentioned in #2.

iv-mexx commented:
> You can run rspec spec/ now. :)

yay! :sparkles:

iv-mexx commented on git-changelog.gemspec:
```diff
@@ -6,10 +6,11 @@ Gem::Specification.new do |s|
   s.license     = 'MIT'
   s.executables = "git-changelog"
   s.add_runtime_dependency 'docopt', '~> 0.5', '>= 0.5.0'
   s.add_runtime_dependency 'rugged'
```
I think we should add `>= 0.23.0` for rugged. The reason for that being in the Gemfile was, that at that time, 0.23 was not yet released and I did not know how specify a git-tag in the gemspec, hence the entry in the Gemfile.

iv-mexx commented on lib/changelog.rb:
```diff
@@ -0,0 +1,111 @@
+    {
+      fixes: @fixes.map(&:note),
+      features: @features.map(&:note),
+      gui: @gui_changes.map(&:note),
```
sneaky breaking change for my fastlane integration :laughing: 
But I like it, its consistent with the tokens.

However, it should be noted as a `refactoring` change!

iv-mexx commented:
Now that there are specs we could also add CI to this project :smiley: 
Also, we'll need some specs for the raw output and then later for traversing the git history, but I'm not yet sure how to do the latter one...

Anyway, nice changes. Pleas add the `refactoring` change to the description. Then we'll see if Github, like Gitlab, takes the description as commit message automatically or if I'll have to add them to a commit message manually...

j4zz commented:
> Now that there are specs we could also add CI to this project :smiley: 

I've pushed a basic `.travis.yml`. After merging, you should be able to easily setup [travis-ci](https://travis-ci.org/).

I've added a refactoring line to the description for the sneaky change :ghost:, and added a line for the travis-ci change as well.

iv-mexx commented:
Nice :+1:

iv-mexx commented:
Ok, in contrast to Gitlab, Github does NOT automatically use the pull request description as commit message for the merge commit, only the title :disappointed:

j4zz commented:
> Ok, in contrast to Gitlab, Github does NOT automatically use the pull request description as commit message for the merge commit, only the title :disappointed:

Maybe add an empty commit on the develop branch with the description as message?

iv-mexx commented:
> Maybe add an empty commit on the develop branch with the description as message?

Yeah, I did that for this PR just now, but its something to keep in mind when using this tool with Github, not as seamless as when working with Gitlab :disappointed:
